src/api/wallet/WalletApiImpl.ts

```diff
@@ -0,0 +1,254 @@
+
+  let blockchainPrompt: BlockchainUpdatePrompt = {
+    account,
+    chainId: +chainId,
+    blockHeader: null,
+  }
+
+  const subscriptionHOC: BlockchainUpdatePromptCallback = callback => {
+    const unsubNetwork = networkUpdate(chainId => {
+      blockchainPrompt = { ...blockchainPrompt, chainId }
+      log('chainId changed:', chainId)
+      callback(blockchainPrompt)
+    })
+    const unsubAccounts = accountsUpdate(([account]) => {
+      blockchainPrompt = { ...blockchainPrompt, account }
+      log('accounts changed:', account)
+      callback(blockchainPrompt)
+    })
+
+    const unsubBlock = blockUpdate(blockHeader => {
+      blockchainPrompt = { ...blockchainPrompt, blockHeader }
+      log('block changed:', blockHeader.number)
+      callback(blockchainPrompt)
+    })
+
+    return (): void => {
+      unsubNetwork()
+      unsubAccounts()
+      unsubBlock()
+    }
+  }
+
+  return subscriptionHOC
+}
+
+// const AUTOCONNECT = process.env.AUTOCONNECT === 'true'
+
+const wcOptions: WalletConnectInits = {
+  package: WalletConnectProvider,
+  options: {
+    // TODO get infuraId from .env
+    infuraId: '8b4d9b4306294d2e92e0775ff1075066',
+  },
+}
+
+/**
+ * Basic implementation of Wallet API
+ */
+export class WalletApiImpl implements WalletApi {
+  private _listeners: ((walletInfo: WalletInfo) => void)[]
+  private _provider: Provider | null
+  private _web3: Web3 | null
+
+  private _unsubscribe: Command = () => {}
+
+  public constructor() {
+    this._listeners = []
+  }
+
+  public isConnected(): boolean {
+    return this._connected
+  }
+
+  public async connect(): Promise<boolean> {
+    const provider = await getProvider(wcOptions)
+
+    if (!provider) return false
+
+    if (isMetamaskProvider(provider)) provider.autoRefreshOnNetworkChange = false
+
+    this._provider = provider
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    this._web3 = new Web3(provider as any)
+    log('[WalletApi] Connected')
+
+    await this._notifyListeners()
+
+    const subscriptions = createSubscriptions(provider)
+
+    const unsubscribeUpdates = subscribeToBlockchainUpdate({ subscriptions, provider, web3: this._web3 })(
+      this._notifyListeners.bind(this),
+    )
+
+    let unsubscribeDisconnect: Command = () => {}
+    if (isWalletConnectSubscriptions(subscriptions)) {
+      unsubscribeDisconnect = subscriptions.onStop(this.disconnect.bind(this))
+    } else if (isMetamaskSubscriptions(subscriptions)) {
+      unsubscribeDisconnect = subscriptions.onAccountsChanged(accounts => {
+        if (accounts.length > 0) return
+        // accounts  = [] when user locks Metamask
+
+        this.disconnect()
+      })
+    }
+
+    this._unsubscribe = (): void => {
+      unsubscribeUpdates()
+      unsubscribeDisconnect()
+    }
+
+    return true
+  }
+
+  public async disconnect(): Promise<void> {
+    if (isWalletConnectProvider(this._provider) && this._connected) await this._provider.close()
```

Comment:
I find this logics strange. 

In my opinion is not right that `Provider` has this signature:
`Provider = WalletConnectProvider | MetamaskProvider`

It should abstract the underling wallet. Otherwise why would we make a library.

I think this line could be just `this._provider.close()` or `if (this._provider.isClosable()) this._provider.close()`

Reply:
As I wrote above it's ill-advised to abstract actual provider integration. Or monkey-patch methods onto it.

Why? Once again, because providers are very different. It's like the difference between HttpConnection and a WebSocket.
Monkey-patching methods or adding a wrapper can interfere with how  3rd-party libraries can handle the provider. Unless you want to wrap in a Proxy, like in `web3 2.0.0`

Reply:
I see you are super strong opinionated about this. I would help for to be on board If I could have reviewed the `'@gnosis.pm/dapp-ui'` PR, but it was to big to be reviewed. So now I'm deriving thoughts that come from that repo and whose discussions would be better than there.

I understand your argument, but I don't suggest `monkey-patch`, I just don't want to see the `if isMetamak do this` logic spread in all the app. If you contain it in the wallet, I'll buy it